toadr3/report.py

```diff
@@ -0,0 +1,122 @@
+import datetime
+
+from .exceptions import SchemaException
+from .interval import Interval
+from .interval_period import IntervalPeriod
+from .payload_descriptor import PayloadDescriptor
+
+
+class ReportData:
+    def __init__(self, data: dict):
+        """Report data for a resource."""
+        if "resourceName" not in data:
+            raise SchemaException("Missing 'resourceName' in report data schema.")
+
+        if "intervals" not in data:
+            raise SchemaException("Missing 'intervals' in report data schema.")
+
+        self._resource_name = data["resourceName"]
+        self._intervals = [Interval(interval) for interval in data["intervals"]]
+        self._interval_period = None
+
+        if "intervalPeriod" in data:
+            self._interval_period = IntervalPeriod(data["intervalPeriod"])
+
+    @property
+    def resource_name(self) -> str:
+        """The name of the resource."""
+        return self._resource_name
+
+    @property
+    def intervals(self) -> list[Interval]:
+        """Applicable intervals for this report and resource."""
+        return self._intervals
+
+    @property
+    def interval_period(self) -> IntervalPeriod | None:
+        """The interval period for this report and resource (if available)."""
+        return self._interval_period
+
+
+class Report:
+    def __init__(self, data: dict):
+        if "objectType" in data and data["objectType"] != "REPORT":
+            raise SchemaException("Expected 'objectType' to be 'REPORT' in report data schema.")
+
+        if "programID" not in data:
+            raise SchemaException("Missing 'programID' in report schema.")
+
+        if "eventID" not in data:
+            raise SchemaException("Missing 'eventID' in report schema.")
+
+        if "clientName" not in data:
+            raise SchemaException("Missing 'clientName' in report schema.")
+
+        if "resources" not in data:
+            raise SchemaException("Missing 'resources' in report schema.")
+
+        self._id = data.get("id", None)
+        self._created = None
+        self._modified = None
+        self._program_id = data["programID"]
+        self._event_id = data["eventID"]
+        self._client_name = data["clientName"]
+        self._report_name = data.get("reportName", None)
+        self._payload_descriptors = None
+        self._resources = [ReportData(resource) for resource in data["resources"]]
+
+        if "createdDateTime" in data:
+            self._created = datetime.datetime.fromisoformat(data["createdDateTime"])
```

Comment:
Is there anything about in the standard about the default timezone in case the timestamps does not contain timezone information?

Reply:
Yes, there is actually a `timeZoneOffset` in the `program` object. Where you can set a duration from UTC. This requires support for negative durations and active use of the `program` object. Will add negative durations, but will postpone the support for time zone offsets.